### PR TITLE
Make MessageQueue growable

### LINKED ISSUE #35653: Change MessageQueue to a page allocator to prevent overflow
**Godot version:**
Master (3cfb67e0f)

**OS/device including version:**
Any

**Issue description:**
As discussed a few times on IRC, and exemplified by many bug reports, our current fixed size for the Message Queue is problematic.

It currently defaults to `1024` (see `memory/limits/message_queue/max_size_kb`), and many users need to increase it to prevent overflow when handling a large amount of messages (which can happen easily in the editor when using e.g. large TileSets #29238 or Skeletons #35652).

As discussed with @reduz, Message Queue should be changed to use a page allocator so that its size can grow as needed, preventing overflows.

Situations which can cause a Message Queue overflow like #29238 or #35652 should however be fixed in the Inspector itself, as even increasing the MQ will just make things functional again, but it will still freeze a long time while populating all those controls.

### PR DESCRIPTION
**Pull request by lawnjelly**
* Uses simple vector resizing (po2)
* Uses pair of read and write buffers
* Monitors queue sizes and periodically shrinks the queue if the reserved capacity is excessive

Co-authored-by: reduz <reduzio@gmail.com>

Fixes #35653
Supersedes #35658
Supersedes #62303

Based on @reduz PR #62303, instead of using a single `LocalVector` buffer, this uses a mirrored pair. This prevents the problems caused by writing to the same `LocalVector` buffer at the same time as reading, which was causing problems.

* This has the advantage of not requiring locks between each Message, making the message queue much more efficient. (I haven't benchmarked just how much faster, but anyone interested could try.)
* The disadvantage is that it potentially can use more memory because of the use of double buffering.
* The memory use is ameliorated because, as reading and writing doesn't happen to the same queue, the peak usage of either queue is lower. So in fact the memory use is likely to be considerably lower than twice.

## Auto-sizing
Although growing with power of 2 resizing is simple, this has the potential to overuse memory because large queue sizes are typically only needed for events such as loading a scene. For this reason, the PR now features a memory monitor that measures the largest used queue size over a period of time, and if the currently used size is significantly below the capacity, the queues are shrunk.

This is done with a large delay to avoid oscillations, and will not shrink the queue below a minimum size of 256Kb (where memory use is less of a concern).

## Notes
* I have tidied up a viable version of the PR because it was on the list of urgent bugs for the beta, and it hadn't been worked on for a while now - reduz may have more pressing areas to work on.
* There are a number of alternative approaches, including using `PagedAllocator` and making local copies of the messages (as I described in https://github.com/godotengine/godot/pull/62303#issuecomment-1186266537 ) . This PR though is probably the most efficient.

### CODE REVIEW

File: core/object/message_queue.cpp [hunk 1]
```diff
@@ -115,26 +129,35 @@ Error MessageQueue::push_callablep(const Callable &p_callable, const Variant **p
...
 
 	int room_needed = sizeof(Message) + sizeof(Variant) * p_argcount;
+	Buffer &buffer = buffers[write_buffer];
 
-	if ((buffer_end + room_needed) >= buffer_size) {
-		print_line("Failed method: " + p_callable);
-		statistics();
-		ERR_FAIL_V_MSG(ERR_OUT_OF_MEMORY, "Message queue out of memory. Try increasing 'memory/limits/message_queue/max_size_kb' in project settings.");
+	if ((buffer.end + room_needed) >= buffer.data.size()) {
+		if ((buffer.end + room_needed) >= max_allowed_buffer_size) {
```
Comment (kleonc) [inline core/object/message_queue.cpp:135]:
```suggestion
	if ((buffer.end + room_needed) > buffer.data.size()) {
		if ((buffer.end + room_needed) > max_allowed_buffer_size) {
```

File: core/object/message_queue.cpp [hunk 2]
```diff
@@ -49,26 +49,31 @@ Error MessageQueue::push_set(ObjectID p_id, const StringName &p_prop, const Vari
...
 
 	uint8_t room_needed = sizeof(Message) + sizeof(Variant);
+	Buffer &buffer = buffers[write_buffer];
 
-	if ((buffer_end + room_needed) >= buffer_size) {
-		String type;
-		if (ObjectDB::get_instance(p_id)) {
-			type = ObjectDB::get_instance(p_id)->get_class();
+	if ((buffer.end + room_needed) >= buffer.data.size()) {
+		if ((buffer.end + room_needed) >= max_allowed_buffer_size) {
```
[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/object/message_queue.cpp]: Good spot, this is also present in #62303 . Have changed. :+1:

[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/object/message_queue.cpp]: Yes, if this PR is approved, then I will backport to 3.x.

**Comment by lawnjelly** [comment]
Note: on rocketchat reduz says he is concerned about keeping memory use low so thinks `PagedAllocator` may be better.

**Comment by lawnjelly** [comment]
In response to @reduz concerns, I have now added auto-shrinkage of the queues (see details in the blurb above). The PR should now as well as being growable and thread efficient, should use less memory than the current master.

**Comment by fire** [comment]
Although there are some merge conflicts what is blocking. this?

**Comment by lawnjelly** [comment]
> Although there are some merge conflicts what is blocking. this?

@reduz wants to try`PagedAllocator`.

Am leaving the PR open though because it seems to work fine, and he does sometimes change his mind later (it's kind of like asking whether you want a different type of burger at restaurant, both will satisfy hunger but you might prefer one over the other). It is already merged in my 3.x fork.

**Comment by lawnjelly** [comment]
Superseded on master by #75940 .

**Pull request closed without merging**